### PR TITLE
8256018: Adler32/CRC32/CRC32C missing reachabilityFence

### PR DESCRIPTION
**Pull request by LanceAndersen**
Hi, 

Please review the fix for JDK-8256018 which addresses the issue that  the update(ByteBuffer)  methods of Adler32, CRC32, and CRC32C should use Reference.reachabilityFence to ensure that direct byte buffer are kept kept alive when they are accessed directly.

The Mach5 jdk-tier1, jdk-tier2, jdk-tier3 as well as the  jck:api/java_util/zip,jck:api/java_util/jar continue to run clean.

Best
Lance

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256018](https://bugs.openjdk.java.net/browse/JDK-8256018): Adler32/CRC32/CRC32C missing reachabilityFence

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 267f2375a76d318b9abd2541bc2e9c5b4e3e70c9

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1149/head:pull/1149`
`$ git checkout pull/1149`

### CODE REVIEW

**Review by naotoj — APPROVED** [review]
Looks good, Lance.

File: src/java.base/share/classes/java/util/zip/Adler32.java [hunk 1]
```diff
@@ -96,8 +97,13 @@ public void update(ByteBuffer buffer) {
         int rem = limit - pos;
         if (rem <= 0)
             return;
-        if (buffer instanceof DirectBuffer) {
-            adler = updateByteBuffer(adler, ((DirectBuffer)buffer).address(), pos, rem);
+        if (buffer.isDirect()) {
+            try {
+                adler = updateByteBuffer(adler, ((DirectBuffer)buffer).address(), pos, rem);
+
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/Adler32.java:103]: spurious blank line here but otherwise looks good.

**Review by AlanBateman — APPROVED** [review]

**Review by naotoj — APPROVED** [review]
Still looks good.

**Comment by simonis** [comment]
I wonder why this change was necessary? Did you see any concrete problems or crashes? If yes, could you please share any details?

The code you've fixed with a `reachabilityFence` looks as follows:
```java
        if (buffer.isDirect()) {
            try {
                crc = updateByteBuffer(crc, ((DirectBuffer)buffer).address(), pos, rem);
            } finally {
                Reference.reachabilityFence(buffer);
            }
        } else if (buffer.hasArray()) {
            ...
        } else {
            ...
        }
        buffer.position(limit); // <-- 'buffer' is still alive here (at least in the bytecode)
```
But as you can see, `buffer` is still alive at the end of the method. According to @iwanowww (see [this mail](http://mail.openjdk.java.net/pipermail/core-libs-dev/2018-February/051312.html) during the ["Reduce Chance Of Mistakenly Early Backing Memory Cleanup" discussion](http://mail.openjdk.java.net/pipermail/core-libs-dev/2018-February/thread.html#51221)) this should suffice because HotSpot computes reachability based on bytecode analysis (rather than analysis of optimized IR). 

Shouldn't this be enough to keep `buffer` alive until `updateByteBuffer()` has finished working on the buffer?

**Pull request closed without merging**